pkg/util/dynclient/dynclient.go

```diff
@@ -0,0 +1,79 @@
+package dynclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+type DynClient interface {
+	GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error)
+}
+
+type dynclient struct {
+	log *logrus.Entry
+
+	dyn dynamic.Interface
+	grs []*restmapper.APIGroupResources
+}
+
+func New(log *logrus.Entry, restconfig *rest.Config) (DynClient, error) {
+	d := &dynclient{
+		log: log,
+	}
+
+	cli, err := discovery.NewDiscoveryClientForConfig(restconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	d.grs, err = restmapper.GetAPIGroupResources(cli)
+	if err != nil {
+		return nil, err
+	}
+
+	d.dyn, err = dynamic.NewForConfig(restconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+func (d *dynclient) GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error) {
+	for _, gr := range d.grs {
+		for version, resources := range gr.VersionedResources {
+			for _, resource := range resources {
```

Comment:
I think you should `continue` if the version doesn't match `gr.Group.PreferredVersion.Version` (double check)